### PR TITLE
[TNL-5941] Adding pagination support

### PR DESCRIPTION
**Pull request by shahbazn**
### Description

Adding paginations support for list videos API

### Sandbox
- [ ] Build a sandbox for your branch and add a link here

### Testing
- [ ] i18n
- [ ] RTL
- [ ] safecommit violation code review process
- [ ] Unit, integration, acceptance tests as appropriate
- [ ] Analytics
- [ ] Performance
- [ ] Database migrations are backwards-compatible

### Reviewers
If you've been tagged for review, please check your corresponding box once you've given the :+1:.
- [ ] Code review: @muhammad-ammar
- [ ] Code review: @mushtaqak
- [ ] Doc Review: @...
- [ ] UX review: @...
- [ ] Accessibility review: @...
- [ ] Product review: @...

### Devops assistance
If it's needed, or even if you suspect it might be needed, do any/all these things as early as possible. Don't surprise devops right before merging!
For example, touching envs/aws.py or or envs/common.py are dead giveaways that you'll need assistance when pushing your changes out.
Things that require devops attention include, but are not limited to:
- configuration changes
- feature flags
- migrations
Ways to get devops attention, pick-and-choose whichever is most appropriate for your specific PR:
- [ ] Mention your concerns to Max in standup, or any other member of devops face-to-face when convinient
- [ ] Tag devops on your PR (@ edx/devops, or @ maxrothman), either for review (more specific to devops = better) or just a general FYI
- [ ] File a support ticket by emailing support@edx.org

FYI: Tag anyone who might be interested in this PR here.

### Post-review
- [ ] Rebase and squash commits

### CODE REVIEW

File: edxval/api.py [hunk 1]
```diff
@@ -317,26 +318,46 @@ def get_url_for_profile(edx_video_id, profile):
...
+    page = paginator.page(page_no)
+    total_count = paginator.count
+    start = paginator.page_range
+    end = paginator.num_pages
+
+    return ({
+        "start": start,
+        "end": end,
+        "total_count": total_count,
+        "page_size": page_size,
```
Comment (muzaffaryousaf) [inline edxval/api.py]: I'd recommend using DRF based serialized data. Can't you get serialized data by using some DRF serializer class ?

File: edxval/api.py [hunk 2]
```diff
@@ -317,26 +318,46 @@ def get_url_for_profile(edx_video_id, profile):
...
+        **kwargs
 ):
     """
     Returns a generator expression that contains the videos found, sorted by
     the given field and direction, with ties broken by edx_video_id to ensure a
     total order.
     """
-    videos = Video.objects.filter(**video_filter)
+    videos_qs = Video.objects.filter(**video_filter)
+    page_no = int(kwargs.pop("page", 1))
```
Comment (muzaffaryousaf) [inline edxval/api.py]: The pagination should be optional based on the given parameter, to make sure that we won't break any other calls to this method.

File: edxval/api.py (hunk 1 above)
Comment (shahbazn) [inline edxval/api.py]: +1 will update

File: edxval/api.py [hunk 3]
```diff
@@ -317,26 +318,46 @@ def get_url_for_profile(edx_video_id, profile):
...
+        sort_dir=SortDirection.asc,
+        **kwargs
 ):
     """
     Returns a generator expression that contains the videos found, sorted by
     the given field and direction, with ties broken by edx_video_id to ensure a
     total order.
     """
-    videos = Video.objects.filter(**video_filter)
+    videos_qs = Video.objects.filter(**video_filter)
```
Comment (muzaffaryousaf) [inline edxval/api.py:330]: I'd leave the name `videos` as it or won't use `qs` at-least.

File: edxval/tests/test_api.py [hunk 4]
```diff
@@ -1081,20 +1081,20 @@ def test_remove_video_for_course(self):
...
+        videos = list(api.get_videos_for_course(self.course_id)["videos"])
         self.assertEqual(len(videos), 0)
 
         # verify that CourseVideo related object exists(soft removal) for removed video
         course_video = CourseVideo.objects.get(course_id=self.course_id, video__edx_video_id=self.edx_video_id)
         self.assertEqual(course_video.is_hidden, True)
 
         # verify that video still exists for other course
-        videos = list(api.get_videos_for_course('other-course'))
+        videos = list(api.get_videos_for_course('other-course')["videos"])
```
Comment (muzaffaryousaf) [inline edxval/tests/test_api.py]: I'd like to see complete test coverage for paginated data.

Comment (muzaffaryousaf) [reply edxval/tests/test_api.py]: To get an rough idea, you can look at notes api test cases.

File: edxval/tests/test_api.py [hunk 5]
```diff
@@ -38,7 +38,7 @@ def check_direction(sort_direction, expected_ids):
             """Assert that the given videos match the expected ids"""
 
             # Make sure it's not just returning the order given
-            actual_videos = api_func(list(reversed(expected_ids)), sort_field, sort_direction)
+            actual_videos = api_func(list(reversed(expected_ids)), sort_field, sort_direction)["videos"]
```
Comment (muzaffaryousaf) [inline edxval/tests/test_api.py]: I'll leave these tests as it and add optional pagination parameter to apply paging and test it through new test cases.

File: edxval/api.py [hunk 6]
```diff
@@ -317,26 +318,46 @@ def get_url_for_profile(edx_video_id, profile):
...
+    page_no = int(kwargs.pop("page", 1))
+    page_size = int(kwargs.pop("page_size", 20))
+
     if sort_field:
         # Refining by edx_video_id ensures a total order
-        videos = videos.order_by(sort_field.value, "edx_video_id")
+        videos_qs = videos_qs.order_by(sort_field.value, "edx_video_id")
         if sort_dir == SortDirection.desc:
-            videos = videos.reverse()
```
Comment (muzaffaryousaf) [inline edxval/api.py]: I'll write a test that'll make sure that videos are reverse and paginated as expected.

**Review by muzaffaryousaf — CHANGES_REQUESTED** [review]
@shahbazn I've left some comments, let me know when you're done with addressing them.

File: edxval/api.py [hunk 7]
```diff
@@ -317,26 +318,40 @@ def get_url_for_profile(edx_video_id, profile):
...
-    return (VideoSerializer(video).data for video in videos)
+            videos_qs = videos_qs.reverse()
+
+    paginated = kwargs.pop("paginated", False)
+    if paginated:
+        page_no = int(kwargs.pop("page", 1))
+        page_size = int(kwargs.pop("page_size", 20))
```
Comment (muzaffaryousaf) [inline edxval/api.py]: Default page size should come from settings.

File: edxval/api.py (hunk 3 above)
Comment (shahbazn) [reply edxval/api.py:330]: its a query set thats why the name to keep it separate from videos below which has the list of videos

File: edxval/serializers.py [hunk 8]
```diff
@@ -205,3 +211,40 @@ def update(self, instance, validated_data):
...
+        Annotate the response with pagination information.
+        """
+        return {
+            'page_size': self.page.paginator.per_page,
+            'total_count': self.page.paginator.count,
+            'total_pages': self.page.paginator.num_pages,
+            'current_page': self.page.number,
+            'sort_field': sort_field.value,
+            'sort_dir': sort_dir.value,
+            'results': data
```
Comment (muzaffaryousaf) [inline edxval/serializers.py]:
```
We use the conventions established by DRF, with some extra fields:
next: A URL to the next page, or null.
previous: A URL to the previous page, or null.
count: The total number of items.
num_pages: The total number of pages.
results: List of results.
The query parameters are as follows:
page - the page of results (zero indexed)
page_size - the number of results to return per page
sort_order - a string indicating the desired order (no direction modifiers)
```

Comment (shahbazn) [reply edxval/serializers.py]: this is using only the fields required by pagination header/footer on the front end side. for example next and previous page links would be redundant if added.

File: edxval/api.py (hunk 6 above)
Comment (shahbazn) [reply edxval/api.py]: added

File: edxval/serializers.py (hunk 8 above)
Comment (muzaffaryousaf) [reply edxval/serializers.py]: and how about other fields total_count, total_pages and current_page ?

Comment (shahbazn) [reply edxval/serializers.py]: total_count used by paging header. total_pages and current_page used by paging footer

**Comment by muzaffaryousaf** [comment]
### Sandbox link --> https://studio-video-paging.sandbox.edx.org/videos/course-v1:edX+DemoX+Demo_Course

File: edxval/api.py (hunk 7 above)
Comment (muzaffaryousaf) [reply edxval/api.py]: Also, i would set the page size to 100 for now.

Comment (muzaffaryousaf) [reply edxval/api.py]: but do it at very end of the all of this work.

File: edxval/serializers.py [hunk 9]
```diff
@@ -205,3 +211,40 @@ def update(self, instance, validated_data):
...
+
+        try:
+            self.page = paginator.page(page_no)
+        except InvalidPage as exc:
+            msg = self.invalid_page_message.format(
+                page_number=page_no, message=six.text_type(exc)
+            )
+            raise NotFound(msg)
+
+        return list(self.page)
```
Comment (muzaffaryousaf) [inline edxval/serializers.py:236]: Why do we need to override `paginate_queryset` method given that we're not doing anything different than the actual method.

File: edxval/serializers.py (hunk 8 above)
Comment (muzaffaryousaf) [reply edxval/serializers.py]: Why redundant fields, i'm not able to see them in response at all.

Comment (muzaffaryousaf) [reply edxval/serializers.py]: Would't be a class like this is enough for edx-val https://github.com/edx/edx-platform/blob/master/openedx/core/lib/api/paginators.py#L10

File: edxval/serializers.py (hunk 9 above)
Comment (shahbazn) [reply edxval/serializers.py:236]: we are. original method requires request object. cannot be used as is.

File: edxval/serializers.py (hunk 8 above)
Comment (shahbazn) [reply edxval/serializers.py]: done

File: edxval/api.py (hunk 7 above)
Comment (shahbazn) [reply edxval/api.py]: done

File: edxval/serializers.py (hunk 9 above)
Comment (shahbazn) [reply edxval/serializers.py:236]: original method requires a request object. and will fail if used as is

**Pull request closed without merging**